#411 Data processing and extraction nodes and script docs (Houdini) (#413)
Inbibo/adonisfx_docs: docs/houdini/tools/ml_data_processing.md

</figure>

2. Connect the simulated skin to process into the first input.

    This skin should contain the simulation data that will be extracted for machine learning. For example, this can be the result of an **AdnSkinMerge** node with the simulation data already present.

3. Connect the rest joints and animated joints.

    These are the same rest and animated joints used to drive the animated skin with a **Bone Deform** node.

4. Connect the capture weights geometry.

    This input must contain the bone capture attributes. It should be a skinned geometry already prepared using the desired skinning approach.

5. Connect the ML joints.

    The ML joints are the joints that will be used as input data during the machine learning data extraction process. In most cases, these are the same as the animated joints, but a dedicated input is provided in case a different joint set is required.

<figure style="width:90%; margin-left:5%" markdown>
  ![AdnMLDataProcessing network view](../images/data_processing_network_view.png)
  <figcaption><b>Figure 2</b>: Network view showing the required connections for the AdnMLDataProcessing HDA.</figcaption>

> </figure>

Reviewer: To keep consistency with the data extraction script page, can we rename the output null nodes to ADN_OUT_SKIN, ADN_OUT_ML_JOINTS?

Reviewer: In fact I think you already use ADN_OUT_SKIN and ADN_OUT_ML_JOINTS in the step 7 below.

Author: OUT_SKIN and OUT_JOINTS is correct, aligned

    We recommend appending **Null** nodes to both outputs and giving them meaningful names, such as **ADN_OUT_SKIN** and **ADN_OUT_ML_JOINTS**. This makes it easier to identify and reference the processed skin and selected ML joints later when configuring the [AdnMLDataExtraction TOP HDA](../tools/data_extraction_tool).

8. Use the outputs in the AdnMLDataExtraction TOP HDA.

    The processed skin and selected ML joints generated by the AdnMLDataProcessing HDA are used as the next step of the ML data extraction workflow. For more information, refer to the [AdnMLDataExtraction TOP HDA](../tools/data_extraction_tool) documentation.

## Parameters

### Joint Selection

| Name | Type | Default | Description |
| :--- | :--- | :------ | :---------- |
| **Load Joint List From File** | Button |  | Loads the joint list from the configuration file of a previous data extraction. This can be used to reuse the same ML joint selection from an existing extracted dataset. |

> | **Joints Group** | Group |  | Group selector used to select the joints that will be used as ML input data during extraction. The selected joints are stored in the **adnMLJoints** point group on the **ADN_OUT_ML_JOINTS** output. |

Reviewer: Do you think we can remove here the "Default" column given that it's empty?

Author: I think we should just keep consistency with what we have in other places

# AdnMLDataProcessing HDA

> 

Reviewer 2: Add "NOTE" box to notify that this feature is available only for ML bundle.

The AdnMLDataProcessing HDA outputs the processed skin data to be extracted from the simulation, together with the selected ML joints. These outputs are then used by the AdnMLDataExtraction TOP HDA as the next step of the workflow.

> [!NOTE]
> The AdnMLDataProcessing HDA is intended to be used before running the [AdnMLDataExtraction TOP HDA](../tools/data_extraction_tool).
> This HDA should not be used in isolation for data extraction. The [AdnMLDataExtraction TOP HDA](../tools/data_extraction_tool) is also required to complete the extraction workflow.

## Requirements

To use the AdnMLDataProcessing HDA, the following inputs must be provided:

- *Input Skin*: Simulated skin geometry to process. This is usually the output of an **AdnSkinMerge** node containing the simulation data.
- *Rest Joints*: KineFX joints in rest pose used by the skinning setup.
- *Anim Joints*: Animated KineFX joints used to drive the animated skin.
- *Capture Weights*: Geometry containing the bone capture attributes required to describe the skinning weights.

> - *ML Joints*: KineFX joints used as input joints for the machine learning data extraction process.

Reviewer 2: ```suggestion
- *ML Joints*: KineFX joints used as driving input joints for the machine learning deformation.
```

Author: Done

The *Anim Joints* and *ML Joints* inputs must contain the following point attributes:

- *name*: Name of each joint.
- *transform*: Transform point attribute of each joint.

> The *Anim Joints* and *ML Joints* inputs are provided separately because the joints used as machine learning inputs do not necessarily need to be the same joints used to deform the geometry. In most cases, the ML joints will be the same as the animated joints, but a dedicated input is available to support workflows where a different joint set is required for data extraction.

Reviewer 2: This concept should go in a note box as it is an advanced use and might confuse users.

```suggestion
The *Anim Joints* and *ML Joints* inputs are provided separately because the joints used as machine learning inputs do not necessarily need to be the same joints used to deform the geometry. In most cases, the ML joints will be the same as the animated joints, but a dedicated input is available to support workflows where a different joint set should be used to drive the ML deformation.
```

Author: Done

    We recommend appending **Null** nodes to both outputs and naming them **OUT_SKIN** and **OUT_JOINTS**. This makes it easier to identify and reference the processed skin and selected ML joints later when configuring the [AdnMLDataExtraction TOP HDA](../tools/data_extraction_tool).

<figure style="width:90%; margin-left:5%" markdown>
  ![AdnMLDataProcessing outputs](../images/data_processing_outputs.png)
  <figcaption><b>Figure 4</b>: Processed outputs generated by the AdnMLDataProcessing HDA. The highlighted wireframe geometry shows the processed skin from the first output, <b>ADN_OUT_SKIN</b>, and the blue joints show the selected ML joints from the second output, <b>ADN_OUT_ML_JOINTS</b>.</figcaption>
</figure>

8. Use the outputs in the AdnMLDataExtraction TOP HDA.

    The processed skin and selected ML joints generated by the AdnMLDataProcessing HDA are used as the next step of the ML data extraction workflow. For more information, refer to the [AdnMLDataExtraction TOP HDA](../tools/data_extraction_tool) documentation.

## Parameters

### Joint Selection

| Name | Type | Default | Description |
| :--- | :--- | :------ | :---------- |

> | *Load Joint List From File* | Button |  | Loads the joint list from the configuration file of a previous data extraction. This can be used to reuse the same ML joint selection from an existing extracted dataset. |

Reviewer 2: ```suggestion
| *Load Joint List From File* | Button |  | Loads the joint list from the configuration file of a previous data extraction. This can be used to reuse the same ML joint selection registered in the `joints.json` file from an existing extracted dataset. |
```

Author: Done

## Result

After cooking the AdnMLDataProcessing HDA, the following outputs are generated:

- **ADN_OUT_SKIN**: The processed skin geometry containing the simulation data to be extracted.
- **ADN_OUT_ML_JOINTS**: The ML joint geometry containing the selected joints for data extraction.

The **ADN_OUT_ML_JOINTS** output includes a point group named *adnMLJoints*. This point group is used by the AdnMLDataExtraction TOP HDA to identify the joints that should be extracted as input data for machine learning.

For clarity, we recommend appending **Null** nodes to these outputs and naming them **OUT_SKIN** and **OUT_JOINTS**. These Null nodes can then be referenced when configuring the data extraction workflow.

The resulting data can then be passed to the [AdnMLDataExtraction TOP HDA](../tools/data_extraction_tool), which gathers the data used for training AdonisML models.

## Limitations

- The pipeline is currently only supported when using the **Linear** skinning method in the **Bone Deform** node.

> - The animated joints and ML joints must contain both the *name* and *transform* point attributes.

Reviewer 3: this is a requirement, not a limitation in my opinion, but you can leave it if you want

Author: Done

> - The ML joints can be different from the animated joints, but they must represent the joint set intended to be used as machine learning input data.

Reviewer 3: this is a requirement, not a limitation in my opinion, but you can leave it if you want

Author: Done

> - The capture weights input must contain valid bone capture attributes.

Reviewer 3: this is a requirement, not a limitation in my opinion, but you can leave it if you want

Author: Done
